File: palindrome/zig/palindrome.zig
const std = @import("std");
const Mutex = std.Thread.Mutex;

fn is_palindrome(number: u32) bool {
    var reversed_number: u32 = 0;
    var original_number = number;

    var n = number;
    while (n != 0) : (n = n / 10) {
        var digit = n % 10;
        reversed_number = reversed_number * 10 + digit;
    }

    return original_number == reversed_number;
}

fn calculate_sum(start: u32, end: u32, sum: *u64, m: *Mutex) void {
    var local_sum: u64 = 0;
    var x = start;
    while (x <= end) {
        if (is_palindrome(x)) {
            local_sum += x;
        }
        x += 1;
    }

    m.lock();
    sum.* += local_sum;
    m.unlock();
}

pub fn main() !void {
    const start = 100_000_000;
    const end = 999_999_999;
    const range = end - start;

    const cores = 4;
    const chunk = range / cores;

    var m = Mutex{};
    var sum: u64 = 0;

    var handles: [cores]std.Thread = undefined;
    var i: u32 = 0;
    while (i < cores) {
        const thread_start = start + (chunk * i);
        const thread_end = if (i == cores - 1) end else (thread_start + chunk - 1);
        //calculate_sum(thread_start, thread_end, &sum, &m);
        handles[i] = try std.Thread.spawn(.{}, calculate_sum, .{ thread_start, thread_end, &sum, &m });
        i += 1;
    }

    for (handles) |handle| {
        handle.join();
    }

    m.lock();
    const stdout = std.io.getStdOut().writer();
    try stdout.print("{d}\n", .{sum});
    m.unlock();
}

File: palindrome/cpp/palindrome.cpp
#include <iostream>
#include <thread>
#include <vector>

bool is_palindrome(int number) {
  int reversed_number = 0;
  int original_number = number;

  int n = number;
  while (n != 0) {
    int digit = n % 10;
    reversed_number = reversed_number * 10 + digit;
    n /= 10;
  }

  return original_number == reversed_number;
}

void calculate_sum(int start, int end, long long &sum) {
  sum = 0;

  for (int x = start; x <= end; x++) {
    if (is_palindrome(x)) {
      sum += x;
    }
  }
}

int main() {
  int start = 100000000;
  int end = 999999999;
  int range = end - start;

  int cores = 4;
  int chunk = range / cores;

  std::vector<std::thread> threads(cores);
  std::vector<long long> results(cores);

  for (int i = 0; i < cores; i++) {
    int t_start = start + (chunk * i);
    int t_end = (i == cores - 1) ? end : t_start + chunk - 1;

    threads[i] =
        std::thread(calculate_sum, t_start, t_end, std::ref(results[i]));
  }

  long long result = 0;

  for (int i = 0; i < cores; i++) {
    threads[i].join();
    result += results[i];
  }

  std::cout << result << std::endl;

  return 0;
}
